components/common-go/namegen/workspaceid.go

```diff
@@ -0,0 +1,484 @@
+// Copyright (c) 2020 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package namegen
+
+import (
+	crand "crypto/rand"
+	"encoding/binary"
+	"log"
+	"math/rand"
+	"regexp"
+	"strings"
+)
+
+// GenerateWorkspaceID generates a new workspace ID by randomly choosing
+func GenerateWorkspaceID() string {
+	return random(colors, 1) + "-" + random(animals, 1) + "-" + random(characters, 8)
+}
+
+var WorkspaceIdPattern = regexp.MustCompile(`^[a-z]{3,12}-[a-z]{2,16}-[a-z0-9]{8}$`)
+
+func random(array []string, length int) string {
+	var result = ""
+	var src cryptoSource
+	rnd := rand.New(src)
```

Comment:
Hm. `cryptoSource` and the usage of `math/rand` seems rather complex at first sight, but maybe I'm missing something.

I thought of:
```
func randomElement(array []string) (string, error) {
    index, err := crand.Int(crand.Reader, len(array))
    if err != nil {
        return nil, err
   }
   
   return array[index], err
}

func randomCharacters(length int) (string, error) {
  bytes := make([]byte, length)
  _, err := crand.Read(bytes)
  if  err != nil {
     return nil, err
  }
  return hex.EncodeToString(bytes), nil

```

Reply:
Not sure I find your proposal more readable. FWIW I consider this code well isolated, dense and tested.

Reply:
> Not sure I find your proposal more readable

Arguments:
1. I had to look 2 times to make sure this this is actually uses "crypto/rand" under the hood: would be nicer to not import "math/rand" at all
2. we're re-implementing `hex.encode`
3. random(array, n) with > n is unnecessary slow: n calls compared to 1

Reply:
Gero has good arguments there :) Personally I also find there are a few levels of abstractions too many in there.
A middle way could be this (in [playground](https://play.golang.org/p/43TrKosDdaM)):
```Go
func GenerateWorkspaceID() (string, error) {
	s1, err := chooseRandomly(colors, 1)
	if err != nil {
		return "", err
	}
	s2, err := chooseRandomly(animals, 1)
	if err != nil {
		return "", err
	}
	s3, err := chooseRandomly(characters, 8)
	if err != nil {
		return "", err
	}

	return strings.Join([]string{s1, s2, s3}, "-"), nil
}

func chooseRandomly(options []string, length int) (res string, err error) {
	l := big.NewInt(int64(len(options)))
	for i := 0; i < length; i++ {
		c, err := rand.Int(rand.Reader, l)
		if err != nil {
			return "", err
		}
		res += options[int(c.Int64())]
	}
	return res, nil
}
```

Hex encoding would use fewer characters and reduce entropy when keepinh the same length.

Reply:
ok, I added your impl. Are we good now?